### PR TITLE
[WIP] Authentication via social networks with schema

### PR DESCRIPTION
**Pull request by ruslankhaertdinov**

### CODE REVIEW

**Comment by ruslankhaertdinov** [comment]
@maximlarionov Этот вариант приближен к алгоритму со схемы, посмотри, пожалуйста.

File: app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  def handle_sign_in
+    user = if auth_verified?
+      VerifiedAuthOrganizer.new(auth).user
+    else
+      UnverifiedAuthOrganizer.new(auth).user
+    end
+    sign_in_and_redirect user, event: :authentication
+  end
```
Comment (maxinspace) [inline app/controllers/omniauth_callbacks_controller.rb:26]: lets's split it into two methods, handler and auth organizer.
First one will handle sign in, and second one will handle 2 cases of auth organizer

``` ruby
def handle_sign_in
  user = auth_organizer.new(auth).user

  sign_in_and_redirect user, event: :authentication
end

def auth_organizer
  auth_verified? ? VerifiedAuthOrganizer : UnverifiedAuthOrganizer
end
```

File: app/interactors/process_user_confirmation.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  attr_reader :user
+  private :user
+
+  def initialize(user)
+    @user = user
+  end
+
+  def call
+    user.confirm
+    # user.update(password: new_password, password_confirmation: new_password)
```
Comment (maxinspace) [inline app/interactors/process_user_confirmation.rb]: excessive comment?

Comment (ruslankhaertdinov) [inline app/interactors/process_user_confirmation.rb]: yep

File: app/controllers/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (ruslankhaertdinov) [inline app/controllers/omniauth_callbacks_controller.rb:26]: I'll split it, thanks

File: app/interactors/unverified_auth_organizer.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  def user
+    if user_found_by_email.present?
+      fail AuthVerificationPolicy::OauthError, "Please, connect your account from profile page."
+    else
+      found_user.send_confirmation_instructions unless found_user.confirmed?
+      found_user
+    end
+  end
```
Comment (maxinspace) [inline app/interactors/unverified_auth_organizer.rb:14]: just don't really like if - else thingies, when they can be split into methods

``` ruby
def user
  ensure_user_existance!

  found_user.send_confirmation_instructions unless found_user.confirmed?
  found_user
end

def ensure_user_existance
  unless user_found_by_email
    fail AuthVerificationPolicy::OauthError, "Please, connect your account from profile page."
  end
end
```

File: spec/features/user/connect_social_account_spec.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      scenario "User connects social account" do
+        expect(page).to have_connected_account("Facebook")
+      end
+    end
+
+    context "user not confirmed" do
+      let!(:user) { create(:user, :from_auth_hashie) }
+
+      scenario "User sees alert" do
+        expect(page).not_to have_css(".js-social-profiles")
```
Comment (maxinspace) [inline spec/features/user/connect_social_account_spec.rb]: checking css in tests ain't cool :3

It's not really readable mostly. Why don't we add another matcher for this case?

File: spec/support/matchers/have_connected_account.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+RSpec::Matchers.define :have_connected_account do |social_profile|
```
Comment (maxinspace) [inline spec/support/matchers/have_connected_account.rb:1]: just add another matcher - have_no_connected_accounts

File: spec/features/user/connect_social_account_spec.rb (hunk 4 above)
Comment (ruslankhaertdinov) [inline spec/features/user/connect_social_account_spec.rb]: In this page can be  2 "Facebook" words - in the prompt section and in the connected profiles list (".js-social-profiles"). And for this reason I check css here. Anyway thanks, I'll try replace it with more robust solution.

File: spec/support/matchers/have_connected_account.rb (hunk 5 above)
Comment (ruslankhaertdinov) [inline spec/support/matchers/have_connected_account.rb:1]: Not sure I understand you :point_up:

**Pull request closed without merging**